DAOS-16473 Support QLC NVMe SSD in control plane and data plane

junzeng1 opened the pull request:
This are the changes for feature of using QLC NVMe SSD to store bulk data. It mainly includes 3 points: 1. support configuring QLC NVMe tier with newly added bulk_data role for both PMem and MD-on-SSD mode, which extend the max tier num to 2 or 4 for each mode respectively; 2. support creating pool with qlc-size parameter to set the size of QLC NVMe SSD in manual mode, and has no impact to the original usage(will set qlc-size to 0 automatically); 3. support display the QLC NVMe SSD and non-QLC NVMe SSD separately for the display of usage and pool related information. Currently, the QLC NVMe tier is identified through the role configured in yml config file. With commits in this PR, you can set a QLC NVMe tier with bulk_data role in yml config file and then start the daos_server to deploy a DAOS storage system with QLC NVMe SSD used to store bulk data. It works well without QLC NVMe tier set(with no bulk_data role set in yml config file), and also support to create pool without using QLC NVMe media when QLC NVMe tier is configured. 

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

junzeng1 committed:
DAOS-16473 control: Support QLC NVMe SSD in control plane

This is the control plane changes for feature of using QLC NVMe SSD to store
bulk data. It mainly includes 3 points: 1, support configuring QLC NVMe tier
with newly added bulk_data role for both PMem and MD-on-SSD mode, which
extend the max tier num to 2 or 4 for each mode respectively; 2, support
creating pool with qlc-size parameter to set the size of QLC NVMe SSD in
manual mode, and has no impact to the original usage; 3, support disaply the
QLC NVMe SSD and non-QLC NVMe SSD seperately for the display of usage and
pool related information. Currently, the QLC NVMe tier is identified
through the role configured in yml config file. With this commit, you can
set a QLC NVMe tier with bulk_data role in yml config file and then start
the daos_server. After that, you can do basic test through dmg command tool.
Note that this commit only implement the control plane, to use the full
feature, will need to wait for another commit for the data plane.

Required-githooks: true

Signed-off-by: Jun Zeng <jun1.zeng@intel.com>

Comphix commented on src/control/server/storage/config.go:
```diff
@@ -431,13 +466,25 @@ func (tcs TierConfigs) validateBdevRoles() error {
 		return nil // MD-on-SSD is not to be enabled
 	}
 
+	// For pmem mode, support data and bulk_data roels set
```
roels --> roles

junzeng1 replied:
ok

kjacque commented on src/control/cmd/dmg/pretty/pool_test.go:
```diff
@@ -432,7 +432,7 @@ Target: type unknown, state up_in
   Total size: 100 GB
   Free: 90 GB
-- Storage tier 2 (unknown):
+- Storage tier 2 (QLC):
```
We should add another tier in these tests to guarantee that unknown tiers print correctly.

junzeng1 replied:
ok, will do it.

kjacque reviewed (commented):
One thing I find myself wondering generally is whether we should present the QLC tier as "QLC" explicitly to the user, as shown here, or whether we should think about a different way to present it. SCM/NVMe tiers are based on which bus the storage is sitting on, with SCM sitting on the memory bus, and NVMe sitting on the PCIe bus. The devices in question may be using any media. Maybe it should simply be considered a second NVMe tier.

@tanabarr has done a lot of work in the MD-on-SSD implementation for the control plane, so I have added him explicitly for review. @mjmac may have some suggestions about how to handle how we present this to users as well.

junzeng1 commented on src/control/lib/control/auto_test.go:
```diff
@@ -1703,10 +1703,11 @@ func TestControl_AutoConfig_genServerConfig(t *testing.T) {
 				MockEngineCfgTmpfs(1, 0, MockBdevTier(1, 3), MockBdevTier(1, 4, 5)),
 			},
-			expErr: errors.New("multiple bdev tiers"),
+			// two bdev Tiers are allowed for pmem mode after QLC supported
+			expErr: errors.New("assigning engine 0 storage bdev tier roles"),
```
actually, I have changed it to 3 bdev tiers for engine 0 in this case already in this PR, so it report error for two many tiers. Please check whether it's ok or not, thanks.

junzeng1 commented on src/engine/init.c:
```diff
@@ -1149,8 +1149,9 @@ parse(int argc, char **argv)
 			rc = arg_strtoul(optarg, &dss_storage_tiers, "\"-T\"");
-			if (dss_storage_tiers < 1 || dss_storage_tiers > 4) {
-				printf("Requires 1 to 4 tiers\n");
+			// add a new tier for QLC NVMe SSD
```
ok, will consider about this. I added a comment here just because this change point is not so clear, I found it not in the coding stage, just found it when I tried to do some test.

junzeng1 commented on src/control/cmd/dmg/pretty/storage.go:
```diff
@@ -99,9 +99,13 @@ func PrintHostStorageUsageMap(hsm control.HostStorageMap, out io.Writer) error {
 	nvmeUsageTitle := "NVMe-Used"
+	qlcTitle := "QLC-Total"
+	qlcFreeTitle := "QLC-Free"
+	qlcUsageTitle := "QLC-Usage"
```
yeah.

junzeng1 commented on src/mgmt/srv_drpc.c:
```diff
@@ -843,7 +843,7 @@ ds_mgmt_drpc_pool_extend(Drpc__Call *drpc_req, Drpc__Response *drpc_resp)
 	uint8_t			*body;
 	size_t			len;
-	uint64_t		scm_bytes, nvme_bytes = 0;
+	uint64_t                 scm_bytes, nvme_bytes = 0, qlc_bytes = 0;
```
ok, will change it.

junzeng1 commented on src/control/server/storage/faults.go:
```diff
@@ -144,10 +144,12 @@ var (
-		"'bdev_roles' are only supported if the scm tier is of class ram so change dcpm tier to ram or "+
-			"remove role assignments from bdev tiers then restart daos_server after updating server "+
-			"config file")
+		"MD-on-SSD roles has been specified in config with scm class set to dcpm or bulk data (QLC) NVMe "+
+			"tier has been configured but roles not set correctly",
+		"change dcpm tier to ram if MD-on-SSD enabled or remove role assignments from bdev tiers if "+
+			"no bulk data (QLC) NVMe tier configured or set 'bdev_role' for data and bulk_data respectively "+
+			"to data NVMe tier and bulk data (QLC) NVMe tier if bulk data (QLC) NVMe tier exist, and then "+
+			"restart daos_server after updating server config file")
```
ok. Will consider to split it to two error codes, and judge each case in src/control/server/storage/config.go .

Comphix commented on src/bio/bio_xstream.c:
```diff
@@ -1658,6 +1658,9 @@ bio_xsctxt_alloc(struct bio_xs_context **pctxt, int tgt_id, bool self_polling)
 		/* No Meta/WAL blobstore if Metadata on SSD is not configured */
 		if (st != SMD_DEV_TYPE_DATA && !bio_nvme_configured(SMD_DEV_TYPE_META))
```
Shall we also update this check, as Metadata on SSD is not configured, it is using PMem, then, Data and Bulk_Data (newly added) is supported, so just check st != TYPE_DATA is missing the TYPE_BULK?

Shall be

(st == TYPE_META || st == TYPE_WAL) && !bio_nvme_configured(TYPE_META)

junzeng1 replied:
seems it should be like this.

Comphix commented on src/control/cmd/dmg/pretty/storage_test.go:
```diff
@@ -934,9 +934,9 @@ Errors:
-Hosts     SCM-Total SCM-Free SCM-Used NVMe-Total NVMe-Free NVMe-Used 
------     --------- -------- -------- ---------- --------- --------- 
-host[1-2] 0 B       0 B      N/A      0 B        0 B       N/A       
+Hosts     SCM-Total SCM-Free SCM-Used NVMe-Total NVMe-Free NVMe-Used QLC-Total QLC-Free QLC-Usage 
```
QLC-Usage --> QLC-Used as others

Comphix commented on src/control/cmd/dmg/pretty/storage_test.go:
```diff
@@ -951,9 +951,9 @@ host[1-2] 0 B       0 B      N/A      0 B        0 B       N/A
-Hosts SCM-Total SCM-Free SCM-Used NVMe-Total NVMe-Free NVMe-Used 
------ --------- -------- -------- ---------- --------- --------- 
-host1 0 B       0 B      N/A      0 B        0 B       N/A       
+Hosts SCM-Total SCM-Free SCM-Used NVMe-Total NVMe-Free NVMe-Used QLC-Total QLC-Free QLC-Usage 
```
QLC-Used

Comphix commented on src/control/cmd/dmg/pretty/storage_test.go:
```diff
@@ -968,9 +968,9 @@ host1 0 B       0 B      N/A      0 B        0 B       N/A
------ --------- -------- -------- ---------- --------- --------- 
-host1 3.0 TB    750 GB   75 %     36 TB      27 TB     25 %      
+Hosts SCM-Total SCM-Free SCM-Used NVMe-Total NVMe-Free NVMe-Used QLC-Total QLC-Free QLC-Usage 
+----- --------- -------- -------- ---------- --------- --------- --------- -------- --------- 
```
QLC-Used

tanabarr reviewed (commented):
I think before we continue reviewing the control-plane side @NiuYawei does this make sense regarding adding an explicit extra tier to the list of "SCM,NVME" in pool query/create schema? Also does the additional role "bulk_data" make sense as applied in this PR? MD-on-SSD branch will have conflicts with this change including updates I am working on for phase-2. TIA

tanabarr commented:
Also, do we really want to add QLC support details for PMem mode? If the additions are only going to be relevant to MD-on-SSD then the changes should be specifically switched based on mode and this PR should be targeted at the feature/vos_on_blob_p2 branch. @gnailzenh @NiuYawei thoughts?

NiuYawei commented:
> I think before we continue reviewing the control-plane side @NiuYawei does this make sense regarding adding an explicit extra tier to the list of "SCM,NVME" in pool query/create schema? Also does the additional role "bulk_data" make sense as applied in this PR? MD-on-SSD branch will have conflicts with this change including updates I am working on for phase-2. TIA

We plan to make QLC as a capacity NVMe tier for bulk data only, it should support both pmem and md-on-ssd mode. See the design at https://daosio.atlassian.net/wiki/spaces/DC/pages/11768692750/QLC+SSD+Support

NiuYawei commented:
> Also, do we really want to add QLC support details for PMem mode? If the additions are only going to be relevant to MD-on-SSD then the changes should be specifically switched based on mode and this PR should be targeted at the feature/vos_on_blob_p2 branch. @gnailzenh @NiuYawei thoughts?

Yes, I think both pmem or md-on-ssd mode should be able to support QLC capacity tier, but I agree with you that it might be more convenient to do the QLC support development on a dedicated feature branch. @gnailzenh , what do you think?

tanabarr commented:
> One thing I find myself wondering generally is whether we should present the QLC tier as "QLC" explicitly to the user, as shown here, or whether we should think about a different way to present it. SCM/NVMe tiers are based on which bus the storage is sitting on, with SCM sitting on the memory bus, and NVMe sitting on the PCIe bus. The devices in question may be using any media. Maybe it should simply be considered a second NVMe tier.
> 
> @tanabarr has done a lot of work in the MD-on-SSD implementation for the control plane, so I have added him explicitly for review. @mjmac may have some suggestions about how to handle how we present this to users as well.

I agree that we should consider it as a second NVMe tier rather than explicitly "QLC". In PMem mode display as "SCM/NVMe 1/NVMe 2" and in md-on-ssd maybe "Meta/Data/Bulk Data"?

kjacque commented:
> Also, do we really want to add QLC support details for PMem mode? If the additions are only going to be relevant to MD-on-SSD then the changes should be specifically switched based on mode and this PR should be targeted at the feature/vos_on_blob_p2 branch.

@tanabarr I agree with this. Unless we have a customer asking for support for this new feature in pmem mode, I think this should be targeted to MD-on-SSD only. Pmem is essentially a legacy feature at this point. New customers aren't likely to be using it, and IMO we should avoid changes to it unless a customer is requesting it.

> I agree that we should consider it as a second NVMe tier rather than explicitly "QLC". In PMem mode display as "SCM/NVMe 1/NVMe 2" and in md-on-ssd maybe "Meta/Data/Bulk Data"?

If we must allow it in pmem mode, maybe "SCM/NVMe/NVMe Bulk"? The names for MD on SSD sound good.

@NiuYawei @gnailzenh In future it would be good to get Control Plane team members a chance to review design docs for a change like this.

junzeng1 committed:
DAOS-16473 vos: Support QLC NVMe SSD in data plane

These are data plane changes for feature of using QLC NVMe SSD to store
bulk data. The main idea is to add a new block allocator for QLC NVMe tier
in vos_pool structure and a new vea_hint_context for bulk data(big size,
e.g., size >= 64k) access in vos_container structure. The QLC NVMe tier is
formated with VOS_BULK_BLK_SZ and data size larger than bulk_data_thresh
(e.g., 64k) from a container will be stored on QLC NVMe tier, including the
data merged through aggregation mechanism in background. And there are few
points to note: 1.this commit depends on commit 499f4c5 and works together
with that one to support the QLC NVMe SSD using feature; 2.for the data
written to QLC NVMe tier, only support async post now, which means that if
an bio contains bio_rsrvd_region to QLC NVMe SSD, than it will be done
synchronously(thus, no UMEM_ACT_CSUM entry for QLC data); 3.the meta blob
header data structure only consider new deployment temporally, not support
the exist pool which created with old version now; 4.the rpc interface is
not compatible with client which build from old version now, for the
DAOS_MEDIA_MAX changed when QLC NVMe SSD added as a new type of media. For
the last two points(3&4), there will be other commit to deal with them, to
make it be compatible with old versions.

Required-githooks: true

Signed-off-by: Jun Zeng <jun1.zeng@intel.com>

junzeng1 commented:
> This are the changes for feature of using QLC NVMe SSD to store bulk data. It mainly includes 3 points: 1. support configuring QLC NVMe tier with newly added bulk_data role for both PMem and MD-on-SSD mode, which extend the max tier num to 2 or 4 for each mode respectively; 2. support creating pool with qlc-size parameter to set the size of QLC NVMe SSD in manual mode, and has no impact to the original usage; 3. support disaply the QLC NVMe SSD and non-QLC NVMe SSD separately for the display of usage and pool related information. Currently, the QLC NVMe tier is identified through the role configured in yml config file. With commits in this PR, you can set a QLC NVMe tier with bulk_data role in yml config file and then start the daos_server to deploy a DAOS storage system with QLC NVMe SSD used to store bulk data. Currently, there are few restrictions for this feature: 1. the meta blob header only consider new deployment temporally, not support the exist pool which created with old version now; 2. the rpc interface is not compatible with old client version now, for the DAOS_MEDIA_MAX changed with QLC NVMe added as an new type of media. And for these two restrictions there will be other commit to deal with them, to be compatible with old versions.
> 
> Required-githooks: true
> 
> ### Before requesting gatekeeper:
> * [ ]  Two review approvals and any prior change requests have been resolved.
> * [ ]  Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
> * [ ]  `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
> * [ ]  Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
> * [ ]  Any tests skipped by the ticket being addressed have been run and passed in the PR.
> 
> ### Gatekeeper:
> * [ ]  You are the appropriate gatekeeper to be landing the patch.
> * [ ]  The PR has 2 reviews by people familiar with the code, including appropriate owners.
> * [ ]  Githooks were used. If not, request that user install them and check copyright dates.
> * [ ]  Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
> * [ ]  All builds have passed.  Check non-required builds for any new compiler warnings.
> * [ ]  Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
> * [ ]  If applicable, the PR has addressed any potential version compatibility issues.
> * [ ]  Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
> * [ ]  Extra checks if forced landing is requested
>   
>   * [ ]  Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
>   * [ ]  No new NLT or valgrind warnings.  Check the classic view.
>   * [ ]  Quick-build or Quick-functional is not used.
> * [ ]  Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

Hello everyone, the data plane implementation has been added to this PR, please help review and give suggestions, thanks.

kjacque reviewed (commented):
Please don't rebase and force-push once your work has begun to be reviewed, or if you really need to (such as for changing base branch), please don't fixup or squash commits. It makes it very hard to tell the changes between versions, especially if old commits are modified.

Instead of rebasing, you can merge the base branch into your PR branch to pick up changes and resolve any conflicts. Commits are cleaned up by the gatekeepers when it lands.

junzeng1 commented:
> Please don't rebase and force-push once your work has begun to be reviewed, or if you really need to (such as for changing base branch), please don't fixup or squash commits. It makes it very hard to tell the changes between versions, especially if old commits are modified.
> 
> Instead of rebasing, you can merge the base branch into your PR branch to pick up changes and resolve any conflicts. Commits are cleaned up by the gatekeepers when it lands.

ok, will do as you mentioned here in the future, thanks for the suggestion. Yesterday, I did rebase the dev branch based on the the master, I thought it may helpful for the reviewer at first. May be I got a wrong feeling, sorry for the inconvenience caused by it. But as you see there are only two commits for this PR, and most of the changes inside it are independent from commits inherited from master branch during the rebasing, so it may still ok to figure out the changes. Thanks.

kjacque commented:
> ok, will do as you mentioned here in the future, thanks for the suggestion. Yesterday, I did rebase the dev branch based on the the master, I thought it may helpful for the reviewer at first. May be I got a wrong feeling, sorry for the inconvenience caused by it. But as you see there are only two commits for this PR, and most of the changes inside it are independent from commits inherited from master branch during the rebasing, so it may still ok to figure out the changes. Thanks.

To be clear, the main concern is that if you make small changes to address comments, and squash/fixup to make them part of your original commits, then we can't easily see the updates. It's unclear to me if that happened here. The Github interface marked some previously-reviewed files as modified, but it's hard to tell if they were meaningful changes, or just a result of rebasing. It's really a problem with how Github displays results from rebasing/force-pushing, but alas, we are all stuck working around it. Thanks.

junzeng1 committed:
DAOS-16473 bio: access exist pool created from version not support QLC

To be compatible with meta data for pool which created with old sotfware
version which not support QLC when doing update from an exist deployment
with previous version.

Required-githooks: true

Signed-off-by: Jun Zeng <jun1.zeng@intel.com>

janekmi commented:
> I think both pmem or md-on-ssd mode should be able to support QLC capacity tier.

@NiuYawei please reconsider whether it is a valuable investment right now. Having it implemented into the PMem mode requires all: coming up with a way to fit it together (will take time for many participants of this discussion), development, writing tests, review, expanding DAOS test infrastructure (already insufficient to cover our needs, @brianjmurrell, @JohnMalmberg), and growing DOAS test matrix (already too big to return a result in a reasonable time; @ryon-jensen).

**I hope we have real customer needs to back up this investment. We are already stretched pretty thin.**

---

> @NiuYawei @gnailzenh In future it would be good to get Control Plane team members a chance to review design docs for a change like this.

@kjacque Amen to this.

---

> If the additions are only going to be relevant to MD-on-SSD then the changes should be specifically switched based on mode and this PR should be targeted at the feature/vos_on_blob_p2 branch.

> MD-on-SSD branch will have conflicts with this change including updates I am working on for phase-2.

@tanabarr From what I know, the development in the MD-on-SSD mode is far more rapid than what is going on in the PMem mode. **So, I would argue that this change ought to target the feature/vos_on_blob_p2 branch regardless**. Otherwise, merging the feature/vos_on_blob_p2 branch to the master will be quite painful, I would expect.

---

> I find myself wondering generally is whether we should present the QLC tier as "QLC" explicitly to the user

> I agree that we should consider it as a second NVMe tier rather than explicitly "QLC". In PMem mode display as "SCM/NVMe 1/NVMe 2" and in md-on-ssd maybe "Meta/Data/Bulk Data"?

> If we must allow it in pmem mode, maybe "SCM/NVMe/NVMe Bulk"? The names for MD on SSD sound good.

@junzeng1 IMHO **the "QLC" word should not appear in the code base under no circumstance**. Otherwise, we tie the DAOS naming convention to a device characteristic and a technology which may change over time. Issues of the QLC technology may be solved in the future, so we may want to use QLC as a different tier as well, or the Octa-level cell will be invented and we would like to use it in the same place as QLC. Either way, it will be awkward. Please don't do it.

My vote is on: "SCM/NVMe/NVMe Bulk" (if we end up adding it to the PMem mode) and "Meta/Data/Bulk Data".

---

This review is massive and since it is already split into two commits **can we review it in two steps as well?** Meaning two separate pull requests:

1. Control plane.
2. Data plane.

It will be easier both for @junzeng1 since they won't need to apply renaming and clean-ups to both commits at the same time, and it may go back and forth as we reasonably can expect, and for the gatekeepers, since they won't need to pick by hand which fix applies to which of the two commits before landing the whole change.

Does it make sense?

junzeng1 committed:
DAOS-16473 pool: Support daos client build from old version

To be compatible with client library which build from old version when
the storage cluster software version is upgraded to new version supported
QLC. And the new daos client is also compatible with cluster with previous
pool rpc version (6).

Required-githooks: true

Signed-off-by: Jun Zeng <jun1.zeng@intel.com>

junzeng1 committed:
DAOS-16473 bio: optimization for nvme_rw completion callback

Remove the allocation for callback arg when supporting QLC, will be
more friendly to performance.

Required-githooks: true

Signed-off-by: Jun Zeng <jun1.zeng@intel.com>

junzeng1 commented:
Have add commits to support exist pool created from old software version which not support the QLC NVMe, and to be compatible with client rpc interface compiled from old software version which not support the QLC NVMe.

janekmi commented on src/control/server/storage/config.go:
```diff
@@ -407,12 +433,18 @@ func (tcs TierConfigs) validateBdevRoles() error {
 		}
 
+		// Disallow having both data and bulk_data on a tier.
+		if hasData && hasBulk {
+			return FaultBdevConfigRolesBulkMixData
+		}
```
```suggestion
		if hasData && hasBulkData {
			return FaultBdevConfigRolesBulkDataMixData
		}
````

janekmi commented on src/control/server/storage/config.go:
```diff
@@ -422,6 +454,9 @@ func (tcs TierConfigs) validateBdevRoles() error {
 		if hasData {
 			data++
 		}
+		if hasBulk {
+			bulk++
+		}
```
```suggestion
		if hasBulkData {
			bulkData++
		}
```

janekmi commented on src/control/fault/code/codes.go:
```diff
@@ -107,6 +107,7 @@ const (
 	BdevConfigControlMetadataNoRoles
 	BdevConfigRolesNoControlMetadata
 	BdevConfigRolesWalDataNoMeta
+	BdevConfigRolesBulkMixData
```
```suggestion
	BdevConfigRolesBulkDataMixData
```

janekmi commented on src/control/server/storage/config_test.go:
```diff
@@ -1,5 +1,5 @@
 //
-// (C) Copyright 2019-2023 Intel Corporation.
+// (C) Copyright 2019-2024 Intel Corporation.
```
You aim at introducing quite a number of new allowed configurations. I would expect a number of new tests here as well. Or am I missing something?

janekmi commented on src/control/server/storage/faults.go:
```diff
@@ -164,6 +166,13 @@ var (
 		"set 'bdev_roles` on all bdev tiers in server config file but avoid the unsupported WAL+Data "+
 			"combination then restart daos_server")
 
+	// FaultBdevConfigRolesBulkMixData indicates an invalid configuration where Bulk_data and Data role are
+	// mix configured on the same NVMe tier.
```
```suggestion
	// FaultBdevConfigRolesBulkDataMixData indicates an invalid configuration where Bulk_data and Data roles are
	// configured on the same NVMe tier.
```

janekmi commented on src/control/server/storage/faults.go:
```diff
@@ -164,6 +166,13 @@ var (
 
+	// FaultBdevConfigRolesBulkMixData indicates an invalid configuration where Bulk_data and Data role are
+	// mix configured on the same NVMe tier.
+	FaultBdevConfigRolesBulkMixData = storageFault(
+		code.BdevConfigRolesBulkMixData,
```
```suggestion
	FaultBdevConfigRolesBulkDataMixData = storageFault(
		code.BdevConfigRolesBulkDataMixData,
```

janekmi commented on src/include/daos_srv/vos_types.h:
```diff
@@ -133,6 +133,9 @@ struct vos_pool_space {
 #define NVME_TOTAL(vps)	((vps)->vps_space.s_total[DAOS_MEDIA_NVME])
 #define NVME_FREE(vps)	((vps)->vps_space.s_free[DAOS_MEDIA_NVME])
 #define NVME_SYS(vps)	((vps)->vps_space_sys[DAOS_MEDIA_NVME])
+#define QLC_TOTAL(vps)  ((vps)->vps_space.s_total[DAOS_MEDIA_QLC])
+#define QLC_FREE(vps)   ((vps)->vps_space.s_free[DAOS_MEDIA_QLC])
+#define QLC_SYS(vps)    ((vps)->vps_space_sys[DAOS_MEDIA_QLC])
```
```suggestion
#define NVME_BULK_TOTAL(vps)  ((vps)->vps_space.s_total[DAOS_MEDIA_NVME_TOTAL])
#define NVME_BULK_FREE(vps)   ((vps)->vps_space.s_free[DAOS_MEDIA_NVME_BULK])
#define NVME_BULK_SYS(vps)    ((vps)->vps_space_sys[DAOS_MEDIA_NVME_BULK])
```

janekmi commented on src/mgmt/srv_drpc.c:
```diff
@@ -696,7 +696,7 @@ ds_mgmt_drpc_pool_evict(Drpc__Call *drpc_req, Drpc__Response *drpc_resp)
 pool_change_target_state(char *id, d_rank_list_t *svc_ranks, size_t n_target_idx,
 			 uint32_t *target_idx, uint32_t rank, pool_comp_state_t state,
-			 size_t scm_size, size_t nvme_size)
+			 size_t scm_size, size_t nvme_size, size_t qlc_size)
```
```suggestion
			 size_t scm_size, size_t nvme_size, size_t nvme_bulk_size)
```

janekmi commented on src/mgmt/srv_drpc.c:
```diff
@@ -765,7 +765,7 @@ ds_mgmt_drpc_pool_exclude(Drpc__Call *drpc_req, Drpc__Response *drpc_resp)
 	rc = pool_change_target_state(req->id, svc_ranks, req->n_target_idx, req->target_idx,
 				      req->rank, PO_COMP_ST_DOWN, 0 /* scm_size */,
-				      0 /* nvme_size */);
+				      0 /* nvme_size */, 0 /* qlc_size */);
```
```suggestion
				      0 /* nvme_size */, 0 /* nvme_bulk_size */);
```

janekmi commented on src/mgmt/srv_drpc.c:
```diff
@@ -814,7 +814,7 @@ ds_mgmt_drpc_pool_drain(Drpc__Call *drpc_req, Drpc__Response *drpc_resp)
 	rc = pool_change_target_state(req->id, svc_ranks, req->n_target_idx, req->target_idx,
 				      req->rank, PO_COMP_ST_DRAIN, 0 /* scm_size */,
-				      0 /* nvme_size */);
+				      0 /* nvme_size */, 0 /* qlc_size */);
```
`nvme_bulk_size`

janekmi commented on src/mgmt/srv_drpc.c:
```diff
@@ -843,7 +843,7 @@ ds_mgmt_drpc_pool_extend(Drpc__Call *drpc_req, Drpc__Response *drpc_resp)
 	uint8_t			*body;
 	size_t			len;
-	uint64_t		scm_bytes, nvme_bytes = 0;
+	uint64_t                 scm_bytes, nvme_bytes = 0, qlc_bytes = 0;
```
`nvme_bulk_bytes`

janekmi commented on src/mgmt/srv_drpc.c:
```diff
@@ -927,6 +931,7 @@ ds_mgmt_drpc_pool_reintegrate(Drpc__Call *drpc_req, Drpc__Response *drpc_resp)
 	size_t				len;
 	uint64_t			scm_bytes;
 	uint64_t			nvme_bytes = 0;
+	uint64_t                         qlc_bytes  = 0;
```
`nvme_bulk_bytes`

janekmi commented on src/mgmt/srv_drpc.c:
```diff
@@ -1740,6 +1749,7 @@ ds_mgmt_drpc_pool_query(Drpc__Call *drpc_req, Drpc__Response *drpc_resp)
 	Mgmt__PoolQueryResp     resp    = MGMT__POOL_QUERY_RESP__INIT;
 	Mgmt__StorageUsageStats scm     = MGMT__STORAGE_USAGE_STATS__INIT;
 	Mgmt__StorageUsageStats nvme    = MGMT__STORAGE_USAGE_STATS__INIT;
+	Mgmt__StorageUsageStats qlc     = MGMT__STORAGE_USAGE_STATS__INIT;
```
`nvme_bulk`

janekmi commented on src/mgmt/srv_internal.h:
```diff
@@ -80,9 +80,10 @@ void ds_mgmt_svc_put(struct mgmt_svc *svc);
+    ds_mgmt_create_pool(uuid_t pool_uuid, const char *group, char *tgt_dev, d_rank_list_t *targets,
 			size_t scm_size, size_t nvme_size, daos_prop_t *prop, d_rank_list_t **svcp,
-			int domains_nr, uint32_t *domains, size_t meta_blob_size);
+			int domains_nr, uint32_t *domains, size_t meta_blob_size, size_t qlc_size);
```
`nvme_bulk_size`

janekmi commented on src/mgmt/srv_internal.h:
```diff
@@ -93,10 +94,10 @@ int ds_mgmt_pool_target_update_state(uuid_t pool_uuid, d_rank_list_t *svc_ranks,
+int
+    ds_mgmt_pool_extend(uuid_t pool_uuid, d_rank_list_t *svc_ranks, d_rank_list_t *rank_list,
+			char *tgt_dev, size_t scm_size, size_t nvme_size, size_t domains_nr,
+			uint32_t *domains, size_t qlc_size);
```
`nvme_bulk_size`

janekmi commented on src/mgmt/srv_pool.c:
```diff
@@ -170,7 +170,7 @@ ds_mgmt_pool_svc_create(uuid_t pool_uuid, int ntargets, const char *group, d_ran
 ds_mgmt_create_pool(uuid_t pool_uuid, const char *group, char *tgt_dev, d_rank_list_t *targets,
 		    size_t scm_size, size_t nvme_size, daos_prop_t *prop, d_rank_list_t **svcp,
-		    int domains_nr, uint32_t *domains, size_t meta_blob_size)
+		    int domains_nr, uint32_t *domains, size_t meta_blob_size, size_t qlc_size)
```
`nvme_bulk_size`

janekmi commented on src/mgmt/srv_pool.c:
```diff
@@ -280,8 +280,8 @@ ds_mgmt_destroy_pool(uuid_t pool_uuid, d_rank_list_t *ranks)
-		    char *tgt_dev,  size_t scm_size, size_t nvme_size, size_t domains_nr,
-		    uint32_t *domains)
+		    char *tgt_dev, size_t scm_size, size_t nvme_size, size_t domains_nr,
+		    uint32_t *domains, size_t qlc_size)
```
`nvme_bulk_size`

janekmi commented on src/mgmt/tests/mocks.c:
```diff
@@ -443,10 +443,9 @@ mock_ds_mgmt_tgt_update_setup(void)
-		    d_rank_list_t *rank_list,
-		    char *tgt_dev,  size_t scm_size, size_t nvme_size,
-		    size_t domains_nr, uint32_t *domains)
+ds_mgmt_pool_extend(uuid_t pool_uuid, d_rank_list_t *svc_ranks, d_rank_list_t *rank_list,
+		    char *tgt_dev, size_t scm_size, size_t nvme_size, size_t domains_nr,
+		    uint32_t *domains, size_t qlc_size)
```
```suggestion
ds_mgmt_pool_extend(uuid_t pool_uuid, d_rank_list_t *svc_ranks, d_rank_list_t *rank_list,
		    char *tgt_dev, size_t scm_size, size_t nvme_size, size_t nvme_bulk_size, size_t domains_nr,
		    uint32_t *domains)
```

As far as I am concern, all the `*_size` arguments should be next to each other.

janekmi commented on src/mgmt/tests/mocks.c:
```diff
@@ -525,7 +524,7 @@ ds_mgmt_group_update_handler(struct mgmt_grp_up_in *in)
 int
 ds_mgmt_create_pool(uuid_t pool_uuid, const char *group, char *tgt_dev, d_rank_list_t *targets,
 		    size_t scm_size, size_t nvme_size, daos_prop_t *prop, d_rank_list_t **svcp,
-		    int domains_nr, uint32_t *domains, size_t meta_blob_size)
+		    int domains_nr, uint32_t *domains, size_t meta_blob_size, size_t qlc_size)
```
```suggestion
ds_mgmt_create_pool(uuid_t pool_uuid, const char *group, char *tgt_dev, d_rank_list_t *targets,
		    size_t scm_size, size_t nvme_size, size_t nvme_bulk_size, daos_prop_t *prop, d_rank_list_t **svcp,
		    int domains_nr, uint32_t *domains, size_t meta_blob_size)
```

The same here.

janekmi reviewed (changes requested):
- control plane (WIP)
- data plane (TBD)

liw commented on src/bio/smd/smd_pool.c:
```diff
@@ -10,21 +10,24 @@
+    "rdb_data",
+    "rdb_meta",
+    "rdb_wal",
+    "rdb_bulk",
```
@janekmi, I think you're right. @NiuYawei, could you confirm, please?

NiuYawei replied:
Yes, QLC won't be used by RDB.

janekmi reviewed (changes requested):
Please take a look also here:

- src/mgmt/tests/srv_drpc_tests.c `init_test_pool_info()`
- src/pool/cli.c `pool_query_target_cb()`
- src/pool/rpc.c `pool_query_reply_to_info_v6()`

My review status:

- control plane (all comments have been published)
- data plane (TBD)

NiuYawei commented:
> > I think both pmem or md-on-ssd mode should be able to support QLC capacity tier.
> 
> @NiuYawei please reconsider whether it is a valuable investment right now. Having it implemented into the PMem mode requires all: coming up with a way to fit it together (will take time for many participants of this discussion), development, writing tests, review, expanding DAOS test infrastructure (already insufficient to cover our needs, @brianjmurrell, @JohnMalmberg), and growing DOAS test matrix (already too big to return a result in a reasonable time; @ryon-jensen).
> 
> **I hope we have real customer needs to back up this investment. We are already stretched pretty thin.**

Yes, I agree that's an issue. @gnailzenh any thoughts?

> 
> > @NiuYawei @gnailzenh In future it would be good to get Control Plane team members a chance to review design docs for a change like this.
> 
> @kjacque Amen to this.
> 

Sure.

kjacque commented on src/control/common/proto/mgmt/pool.pb.go:
```diff
@@ -6,8 +6,8 @@
 // versions:
-// 	protoc-gen-go v1.31.0
-// 	protoc        v3.5.0
+// 	protoc-gen-go v1.34.2
+// 	protoc        v3.11.4
```
It's usually not a big deal to have some flux between generator versions as long as they are reasonably close to each other.  As long as it's not making big changes in formatting of the generated files, I don't have strong feelings about it. It looks like the changes applied cleanly in this case.

junzeng1 commented:
> > I think both pmem or md-on-ssd mode should be able to support QLC capacity tier.
> 
> @NiuYawei please reconsider whether it is a valuable investment right now. Having it implemented into the PMem mode requires all: coming up with a way to fit it together (will take time for many participants of this discussion), development, writing tests, review, expanding DAOS test infrastructure (already insufficient to cover our needs, @brianjmurrell, @JohnMalmberg), and growing DOAS test matrix (already too big to return a result in a reasonable time; @ryon-jensen).
> 
> **I hope we have real customer needs to back up this investment. We are already stretched pretty thin.**
> 
> > @NiuYawei @gnailzenh In future it would be good to get Control Plane team members a chance to review design docs for a change like this.
> 
> @kjacque Amen to this.
> 
> > If the additions are only going to be relevant to MD-on-SSD then the changes should be specifically switched based on mode and this PR should be targeted at the feature/vos_on_blob_p2 branch.
> 
> > MD-on-SSD branch will have conflicts with this change including updates I am working on for phase-2.
> 
> @tanabarr From what I know, the development in the MD-on-SSD mode is far more rapid than what is going on in the PMem mode. **So, I would argue that this change ought to target the feature/vos_on_blob_p2 branch regardless**. Otherwise, merging the feature/vos_on_blob_p2 branch to the master will be quite painful, I would expect.
> 
> > I find myself wondering generally is whether we should present the QLC tier as "QLC" explicitly to the user
> 
> > I agree that we should consider it as a second NVMe tier rather than explicitly "QLC". In PMem mode display as "SCM/NVMe 1/NVMe 2" and in md-on-ssd maybe "Meta/Data/Bulk Data"?
> 
> > If we must allow it in pmem mode, maybe "SCM/NVMe/NVMe Bulk"? The names for MD on SSD sound good.
> 
> @junzeng1 IMHO **the "QLC" word should not appear in the code base under no circumstance**. Otherwise, we tie the DAOS naming convention to a device characteristic and a technology which may change over time. Issues of the QLC technology may be solved in the future, so we may want to use QLC as a different tier as well, or the Octa-level cell will be invented and we would like to use it in the same place as QLC. Either way, it will be awkward. Please don't do it.
> 
> My vote is on: "SCM/NVMe/NVMe Bulk" (if we end up adding it to the PMem mode) and "Meta/Data/Bulk Data".
> 
> This review is massive and since it is already split into two commits **can we review it in two steps as well?** Meaning two separate pull requests:
> 
> 1. Control plane.
> 2. Data plane.
> 
> It will be easier both for @junzeng1 since they won't need to apply renaming and clean-ups to both commits at the same time, and it may go back and forth as we reasonably can expect, and for the gatekeepers, since they won't need to pick by hand which fix applies to which of the two commits before landing the whole change.
> 
> Does it make sense?

Yes, will try to split this PR to serval small PRs and try to get them reviewed and merged gradually later. For the tier name, will change  it as all the discussion here.

junzeng1 commented on src/control/server/engine/utils.go:
```diff
@@ -23,6 +23,8 @@ const (
 	NvmeMinBytesPerTarget = 1 * humanize.GiByte
 	// ScmMinBytesPerTarget is min SCM pool allocation per target
 	ScmMinBytesPerTarget = 16 * humanize.MiByte
+	// QlcMinBytesPerTarget is min QLC NVMe pool allocation per target
+	QlcMinBytesPerTarget = 1 * humanize.GiByte
```
only works when bulk_data tier is requested during the creation of pool when bulk_data role set in yml configure file.

junzeng1 commented on src/control/server/storage/bdev.go:
```diff
@@ -412,27 +414,60 @@ func (ncs NvmeControllers) Capacity() (tb uint64) {
 }
 
-// Free returns the cumulative available bytes of all blobstore clusters.
+// Free returns the cumulative available bytes of all blobstore clusters which are not Bulk role.
```
to display capacity of common NVMe and QLC NVMe separately.

junzeng1 commented on src/control/cmd/dmg/pool.go:
```diff
@@ -232,6 +236,10 @@ func (cmd *PoolCreateCmd) storageAutoTotal(req *control.PoolCreateReq) error {
 		return errIncompatFlags("nranks", "ranks")
 	}
 
+	if cmd.QlcSize.IsSet() {
+		return errIncompatFlags("total", "qlc-size")
+	}
```
It's in storageAutoTotal. Currently, not support set each tier size automatically when NvmeBulkSize is set.

junzeng1 commented on src/control/cmd/dmg/pretty/pool.go:
```diff
@@ -98,9 +100,12 @@ func PrintPoolCreateResponse(pcr *control.PoolCreateResp, out io.Writer, opts ..
 	tierName := "SCM"
 	for tierIdx, tierRatio := range tierRatios {
-		if tierIdx > 0 {
+		if tierIdx == 1 {
 			title += ","
 			tierName = "NVMe"
+		} else if tierIdx == 2 {
+			title += ","
+			tierName = "QLC"
 		}
```
Currently, for NVME_BULK only support set it's size with specified parameter in manual mode. So there is only one tier ratio now.
